bigtable/client/row_range.cc

```diff
@@ -0,0 +1,446 @@
+    case btproto::RowRange::END_KEY_NOT_SET:
+      type = KeyType::NOT_SET;
+      break;
+  }
+  return type;
+}
+
+/**
+ * Get the start key: if neither field is set, interpreted as
+ * empty key, inclusive.
+ */
+std::string RowRange::GetStartKey() const {
+  std::string result = "";
+
+  switch (range_.start_key_case()) {
+    case btproto::RowRange::kStartKeyOpen:
+      result = range_.start_key_open();
+      break;
+    case btproto::RowRange::kStartKeyClosed:
+      result = range_.start_key_closed();
+      break;
+    case btproto::RowRange::START_KEY_NOT_SET:
+      break;
+  }
+  return result;
+}
+
+/**
+ * Get the end key: if neither filed is set interpreted as
+ * infinite key, exclusive.
+ */
+std::string RowRange::GetEndKey() const {
+  std::string result = "";
+
+  switch (range_.end_key_case()) {
+    case btproto::RowRange::kEndKeyOpen:
+      result = range_.end_key_open();
+      break;
+    case btproto::RowRange::kEndKeyClosed:
+      result = range_.end_key_closed();
+      break;
+    // Same as btproto::RowRange::START_KEY_NOT_SET.
+    case btproto::RowRange::END_KEY_NOT_SET:
+      break;
+  }
+  return result;
+}
+
+/**
+ * Finalize the end points for intersection.
+ * position : gives the point position (start or end).
+ */
+void RowRange::FixKeyPointAndTypeIntersect(RowRange::PositionType position,
+                                           std::string& result_key,
+                                           RowRange::KeyType& result_type,
+                                           std::string rhs_key,
+                                           RowRange::KeyType rhs_type) {
+  // Pick the max value for start point, and min value for the end.
+  bool comparison = (position == PositionType::START ? result_key < rhs_key
+                                                     : result_key > rhs_key);
+  if (comparison) {
+    result_key = rhs_key;
+    result_type = rhs_type;
+  } else if (result_key == rhs_key && KeyType::OPEN == rhs_type) {
+    // prefer OPEN type in case of a tie.
+    result_type = rhs_type;
+  }
+}
+
+/**
+ * Initialize the protocol Buffer within the RowRange.
+ */
+void RowRange::SetProtoBufRangePoint(RowRange::PositionType position,
+                                     RowRange& row_range, std::string point_key,
+                                     RowRange::KeyType point_type) {
+  if (position == PositionType::START) {
+    if (point_type == KeyType::OPEN)
+      row_range.range_.set_start_key_open(point_key);
+    else if (point_type == KeyType::CLOSED)
+      row_range.range_.set_start_key_closed(point_key);
+  } else {  // End point.
+    if (point_type == KeyType::OPEN)
+      row_range.range_.set_end_key_open(point_key);
+    else if (point_type == KeyType::CLOSED)
+      row_range.range_.set_end_key_closed(point_key);
+  }
+}
+
+std::string RowRange::debug_string() const {
+  KeyType start_type{GetStartKeyType()}, end_type{GetEndKeyType()};
+
+  std::string result = "";
```

Comment:
I think you can do this with the enums provided by the proto.

Reply:
The start and end points use two different enum classes start and end key.  I converted them to a single enum to simplify the "intersect" code, and to use the same code to 
 compute the start and end  type of the range. 
start type (e.g., open or closed) of the intersect output range could come from either start or end type of the input ranges.

Reply:
Yes, for this method, I can do directly with the enums from the proto. Thanks.